### PR TITLE
Multiple Improvements (loglevel, kernel module, change fstype detection, container)

### PR DESCRIPTION
**Pull request by jeremydenoun**
I propose this PR with:
- add cli options to override configuration loglevel
- add new kernel module detection (enable & listing)
- change way to detect if file system type is disabled
- add global IS_CONTAINER variable
- clean previous commit to respect same declaration function coding style

### CODE REVIEW

File: lib/utils.sh [hunk 1]
```diff
@@ -344,6 +344,44 @@ is_kernel_option_enabled() {
...
 }
 
+# Detect if module $1 is enabled, can be filtered with $2
+is_kernel_module_enabled() {
+    local MODULE_NAME=$1
+    FNRET=128
+    local module
+
+    if [ $# -eq 2 ]; then
+        module="$($SUDO_CMD modprobe -n -v "$MODULE_NAME" 2>/dev/null | grep -E "$2" | xargs)"
```
Comment (thibaultserti) [inline lib/utils.sh:354]: neither `modprobe` nor `lsmod` work on a monolithic kernel, that's why we prefer to look at /proc/config.gz instead.
This command will fail on a monolithic kernel.
Do you see any benefit doing the way you propose ?

File: bin/hardening/1.1.1.1_disable_freevxfs.sh [hunk 2]
```diff
@@ -17,28 +17,43 @@ HARDENING_LEVEL=2
 # shellcheck disable=2034
 DESCRIPTION="Disable mounting of freevxfs filesystems."
 
-# Note: we check /proc/config.gz to be compliant with both monolithic and modular kernels
-
-KERNEL_OPTION="CONFIG_VXFS_FS"
-MODULE_NAME="freevxfs"
+FSTYPE_NAME="freevxfs"
```
Comment (thibaultserti) [inline bin/hardening/1.1.1.1_disable_freevxfs.sh:20]: You forgot to change 1.1.1.23, 3.4.1, 3.4.2, 3.4.3, 3.4.4 that use the same function `is_kernel_option_enabled`

File: bin/hardening/1.1.1.1_disable_freevxfs.sh [hunk 3]
```diff
@@ -17,28 +17,43 @@ HARDENING_LEVEL=2
...
 
 # This function will be called if the script status is on enabled / audit mode
 audit() {
-    is_kernel_option_enabled "$KERNEL_OPTION" "$MODULE_NAME"
-    if [ "$FNRET" = 0 ]; then # 0 means true in bash, so it IS activated
-        crit "$KERNEL_OPTION is enabled!"
+    if [ "$IS_CONTAINER" -eq 1 ]; then
+        # In an unprivileged container, the kernel modules are host dependent, so you should consider enforcing it
+        ok "Container detected, consider host enforcing!"
```
Comment (thibaultserti) [inline bin/hardening/1.1.1.1_disable_freevxfs.sh:26]: change with:
"Container detected, consider host enforcing or disable this check!"

File: bin/hardening.sh [hunk 4]
```diff
@@ -26,6 +26,7 @@ ALLOW_SERVICE_LIST=0
 SET_HARDENING_LEVEL=0
 SUDO_MODE=''
 BATCH_MODE=''
+ASK_LOGLEVEL=''
```
Comment (thibaultserti) [inline bin/hardening.sh:29]: Ok for LOGLEVEL it's much better this way !

**Review by thibaultserti — CHANGES_REQUESTED** [review]

File: lib/common.sh [hunk 5]
```diff
@@ -113,11 +113,36 @@ sudo_wrapper() {
     fi
 }
 
+#
+# detect if container based on cgroup detection
```
Comment (thibaultserti) [inline lib/common.sh:117]: Ok for this implementation !

File: bin/hardening.sh [hunk 6]
```diff
@@ -98,6 +99,10 @@ OPTIONS:
         the '-n' option instructs sudo not to prompt for a password.
         Finally note that '--sudo' mode only works for audit mode.
 
+    --set-log-level <level>
```
Comment (thibaultserti) [inline bin/hardening.sh:102]: Need to be added to MANUAL.md and README.md
(yes it's not really efficient, you must write it in 3 differents places :disappointed: )

File: bin/hardening/1.1.1.1_disable_freevxfs.sh (hunk 2 above)
Comment (jeremydenoun) [reply bin/hardening/1.1.1.1_disable_freevxfs.sh:20]: You have right, I propose to finish discussion on remaining point to decide which kind of final fix should be done

File: lib/utils.sh (hunk 1 above)
Comment (jeremydenoun) [reply lib/utils.sh:354]: If you check CIS debian recommendations (https://www.cisecurity.org/ - CIS_Debian_Family_Linux_Benchmark_v1.0.0) you can find them preferred benchmark method based on kernel module testing and listing but you are right on monolithic kernel this can be an error. I think your actual approach based on file detection in `/lib/modules/$(uname -r)` is problematic because you can have a module overridden (on startup with install or after boot) and be a secure behavior. This can be more secure to check real module dry-run loading too.

Maybe I can merge both approach with 2 options : 

1. remove module detection from `is_kernel_option_enabled` and check `is_kernel_option_enabled` and mine `is_kernel_module_enabled`
2. use only `is_kernel_option_enabled` add `"install /bin/(true|false)"` detection and change detection (modprobe first, file detection fallback)

What do you think ?

Comment (thibaultserti) [reply lib/utils.sh:354]: In fact I think that the second option is the best yes :
* first use your implementation to check with modprobe or lsmod
* in case of failure you'll get something like that
```console
# lsmod
libkmod: ERROR ../libkmod/libkmod-module.c:1657 kmod_module_new_from_loaded: could not open /proc/modules: No such file or directory
Error: could not get list of modules: No such file or directory
```
and you can fallback to the method which check  `/proc/config.gz`

**Review by thibaultserti — COMMENTED** [review]
Ok for this changes, waiting for the requested changes on `is_kernel_option_enabled`

File: lib/utils.sh (hunk 1 above)
Comment (jeremydenoun) [reply lib/utils.sh:354]: do you have on hand the `lsmod` return code on this case ? this is maybe the best way to detect it

Comment (thibaultserti) [reply lib/utils.sh:354]: The return code is 1. The error message is the one I wrote upper.

Comment (jeremydenoun) [reply lib/utils.sh:354]: thx

File: bin/hardening/1.1.1.1_disable_freevxfs.sh (hunk 2 above)
Comment (jeremydenoun) [reply bin/hardening/1.1.1.1_disable_freevxfs.sh]: fixed in 678425dab5d62da84f095c4bbe62abf8646ea163 I update "tests set" to be consistent (not checked on docker)

File: lib/utils.sh (hunk 1 above)
Comment (jeremydenoun) [reply lib/utils.sh]: implement in 678425dab5d62da84f095c4bbe62abf8646ea163 please note this implementation can impact an inconsistency with sudo flag (example if you don't have right to lsmod as a normal user we will fallback to kernel options detection (mainly false-positive) and with sudo we will check with modprobe that can bring an inconsistency)

File: lib/utils.sh [hunk 7]
```diff
@@ -295,52 +295,68 @@ is_service_enabled() {
...
-        debug "Kernel option $KERNEL_OPTION disabled"
-        FNRET=1
-    else
-        debug "Kernel option $KERNEL_OPTION not found"
-        FNRET=2 # Not found
+    if $SUDO_CMD lsmod >/dev/null 2>&1; then
+        IS_MONOLITHIC_KERNEL=0
     fi
 
-    if $SUDO_CMD [ "$FNRET" -ne 0 ] && [ -n "$MODULE_NAME" ] && [ -d "/lib/modules/$(uname -r)" ]; then
```
Comment (thibaultserti) [inline lib/utils.sh:326]: Why did you remove this part ? I think it's still relevant (in both cases)

File: lib/utils.sh [hunk 8]
```diff
@@ -295,52 +295,68 @@ is_service_enabled() {
...
-    if [ "x$ANSWER" = "xy" ]; then
-        debug "Kernel option $KERNEL_OPTION enabled"
-        FNRET=0
-    elif [ "x$ANSWER" = "xn" ]; then
-        debug "Kernel option $KERNEL_OPTION disabled"
-        FNRET=1
-    else
-        debug "Kernel option $KERNEL_OPTION not found"
-        FNRET=2 # Not found
+    if $SUDO_CMD lsmod >/dev/null 2>&1; then
```
Comment (thibaultserti) [inline lib/utils.sh:311]: Maybe add a debug log here

**Review by thibaultserti — CHANGES_REQUESTED** [review]

File: lib/utils.sh (hunk 7 above)
Comment (jeremydenoun) [reply lib/utils.sh:326]: Regarding CIS recommendations if you override module (with install /bin/(true|false)) and lsmod return nothing, this is considered as secure, with your version we add kernel option detection for monolithic kernel.

If you consider we should detect if modules file exists, check if blacklist or override, this is redundant if `lsmod/modprobe` work but maybe you have case (example WSL) where a monolithic kernel don't have a /proc/modules and can use dynamic modules so I propose to restore check only if `lsmod/modprobe` doesn't work, would that be suitable ? (if yes check my last commit)

PS: regarding modprobe.d manpage (https://linux.die.net/man/5/modprobe.d) if I restore the check I should change blacklist/install `find` command to only check `.conf` file (`All files underneath the /etc/modprobe.d directory which end with the .conf extension specify those options as required`)

File: lib/utils.sh (hunk 8 above)
Comment (jeremydenoun) [reply lib/utils.sh:312]: fixed 557c46f

File: lib/utils.sh (hunk 7 above)
Comment (thibaultserti) [reply lib/utils.sh:326]: > If you consider we should detect if modules file exists, check if blacklist or override, this is redundant if lsmod/modprobe work but maybe you have case (example WSL) where a monolithic kernel don't have a /proc/modules and can use dynamic modules so I propose to restore check only if lsmod/modprobe doesn't work, would that be suitable ? (if yes check my last commit)
Yes that was what I meant
Ok for the  èfind`

**Review by thibaultserti — APPROVED** [review]
